fastlane/fastlane
# [fastlane] load additional .env files from within CLIToolsDistributor

### Checklist

### Motivation and Context

Closes #16077 

As I raised in the linked ticket, currently the additional .env files specified with `--env` are only supported when running a lane directly. This means that you cannot run any specific tool or action and be able to pass the `--env` argument without wrapping it in a lane.

This behaviour is odd, since the **.env** and **.env.default** files however are supported.

I work on a large project that has five different configurations that we can build and release that all use different code signing and bundle identifiers. We use XcodeGen to generate the project and rely heavily on injecting environment variables to define the differing build settings.

To support the above project, we use custom **.env** files to overload the environment variables for the given configuration... For example, `fastlane build --env internal`. This is great, however to share the environment with XcodeGen, we have a custom action and to avoid repetition between our CI setup and local setups, we invoke this locally to generate our project as well. 

Since `fastlane run my_action` doesn't support the `--env` argument however, we're unable to generate projects for different environments locally without wrapping the action in a dummy lane like such:

```ruby
lane :generate do
    my_action
end
```

### Description

The changes in this PR intend of supporting the use of the --env argument universally within the Fastlane cli however this was a bit trickier than I was initially expecting... `CLIToolsDistributor` is a wrapper that basically works out which Commander interface needs to be invoked and as a result, it doesn't have access to any of Commanders nice argument parsing. 

I was able to work around this using the simple `env_cl_param` lambda in the `load_dot_env` method that I've introduced on `CLIToolsDistributor` however I don't think its very elegant and would appreciate some input 🙏 

The general gist however is that we strip out `--env blah` from `ARGV` here before invoking any of the defined tools, main command interface or credentials manager. 

To keep the visibility of this option, I've then gone through every `CommandsGenerator` class and listed `--env` as a global_option, kind of following how `--verbose` seems to be handled. I'm not sure if this is the right approach though, since there is a bit of a disconnect between where we parse `--env` and where its defined on the command interface... Will appreciate some input here also.

~Additionally I've left three TODO's in the code that I need some help answering.~ (answered now, thanks @joshdholtz) 

### Testing Steps

Amongst the changes, I've both removed old tests and introduced new ones... I added a comprehensive suite of tests to `DotenvHelper` since its functionality was previously being tested as part of the *fastlane/spec/commands_generator_spec.rb** and **fastlane/spec/lane_manager.rb** however since we have moved the loading up a level, this is no longer the correct place to test. 

I also added tests to **cli_tools_distributor_spec.rb** however instead of checking that the `ENV` object is loaded as expected, the tests focus on checking that the `DotenvHelper` is invoked correctly.

I realise that these changes have kind of focused on testing individual components now and I haven't really added any kind of test to verify that the entire process works as a whole since I wasn't really sure how to approach it.. I was thinking that maybe it'll be good to test somewhere that invoking `fastlane test_lane` would result in an output that prints `ENV['SOME_VAR']` where the value is expected based on the **.env** setup... I'm not sure if this is a bit much or not so again I need some guidance here 🙇 


Thanks again for the great tooling, it's really helping our team and I'm enjoying pushing it to the limits and getting stuck in trying to improve things for everybody 🚀
